src/NuGet.Services.EndToEnd/Support/Clients/V2V3SearchClient.cs

```diff
@@ -323,37 +326,93 @@ private async Task<SearchServiceProperties> GetSearchServiceFromAzureAsync(Servi
             var complete = false;
-            do
+            var cached = false;
+            lock (_pollCacheLock)
+            {
+                if (_pollCache.TryGetValue(url, out response))
+                {
+                    complete = isComplete(response);
+                    cached = complete;
+
+                    // If the cached response is not considered complete, remove it from the cache.
+                    if (!complete)
+                    {
+                        logger.WriteLine($"The cached response for {id} {version} is no longer valid.");
+                        _pollCache.Remove(url);
+                    }
+                    else
+                    {
+                        logger.WriteLine($"The cached response for {id} {version} will be used.");
+                    }
+                }
+            }
+
+            var additionalPolling = _testSettings.SearchServiceConfiguration.AdditionalPollingDuration;
+            var duration = Stopwatch.StartNew();
+            var sinceFirstComplete = new Stopwatch();
+            var attempts = 0;
+            while (!cached                                                 // Keep polling if the response was not cached.
+                   && duration.Elapsed < TestData.SearchWaitDuration       // Keep polling if duration is less than the max.
+                   && (!complete                                           // Keep polling if we are not yet complete...
+                       || sinceFirstComplete.Elapsed < additionalPolling)) // ... or if we haven't polled for the additional duration.
             {
-                var response = await _httpClient.GetJsonAsync<V2SearchResponse>(
+                if (attempts > 0)
+                {
+                    await Task.Delay(TestData.V3SleepDuration);
+                }
+
+                attempts++;
+
+                response = await _httpClient.GetJsonAsync<V2SearchResponse>(
                     url,
                     allowNotFound: false,
                     logResponseBody: false,
                     logger: logger);
 
                 complete = isComplete(response);
 
-                if (!complete && duration.Elapsed + TestData.V3SleepDuration < TestData.SearchWaitDuration)
+                if (complete)
                 {
-                    await Task.Delay(TestData.V3SleepDuration);
+                    if (additionalPolling > TimeSpan.Zero && !sinceFirstComplete.IsRunning)
+                    {
+                        logger.WriteLine($"Package {id} {version} is now complete. Polling for an additional {additionalPolling}.");
+                        sinceFirstComplete.Start();
+                    }
+                }
+                else
+                {
+                    if (sinceFirstComplete.IsRunning)
+                    {
+                        logger.WriteLine($"Package {id} {version} is no longer complete. Restarting the timer.");
```

Comment:
>Restarting [](start = 89, length = 10)

I'd suggest using the word `Resetting` instead of `Restarting`, considering that these operations mean different things on a C# stopwatch

Reply:
Yeah, good point.

Reply:
Done.